dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/export/trackedentity/HibernateTrackedEntityStore.java

```diff
@@ -632,6 +614,40 @@ private String getFromSubQueryJoinOrgUnitConditions(TrackedEntityQueryParams par
     return orgUnits.toString();
   }
 
+  /**
+   * In case of a regular user, generates a subquery to find the TE/program pair owner. It will
+   * return the TE registering org unit if none found. In case of a superuser, will return the TE
+   * registering org unit directly.
+   */
+  private String getOwnerOrgUnitsForAccessibleTrackedEntityPrograms(
+      TrackedEntityQueryParams params) {
+    if (skipOwnershipCheck(params)) {
+      return "(TE.organisationunitid)";
+    }
+
+    String sql =
+        "(select distinct coalesce (tpo.organisationunitid, te.organisationunitid) "
+            + "from trackedentitytype tet "
+            + "left join trackedentityprogramowner tpo on tpo.trackedentityid = te.trackedentityid "
+            + "left join program p on p.programid = tpo.programid "
+            + "and tet.trackedentitytypeid = te.trackedentitytypeid "
+            + "and p.trackedentitytypeid = te.trackedentitytypeid "
+            + "and p.programid in ("
+            + getCommaDelimitedString(getIdentifiers(params.getPrograms()))
+            + ")";
+
+    if (params.hasTrackedEntityType()) {
```

Comment:
this query doesn't look good to me. the access plan is not promising see https://explain.dalibo.com/plan/2aac13cc0c98a97b
Also, i don't see why you need to start from the `trackedentytype` table if you already know it from the program and the te. I'll see how to improve the query or you can check it out as well

Reply:
After the refactor, the query plan looks better: https://explain.dalibo.com/plan/5g16549b8b050gbb

Reply:
@muilpp ask @radnov to set up an instance with a performance database, use the https://github.com/dhis2/performance-tests-gatling to see performances. You can compare the result with the https://test.performance.dhis2.org/dev instance I've used recently for https://github.com/dhis2/dhis2-core/pull/16965